### PR TITLE
feat: support prepare pushdown

### LINKED ISSUE #51: PREPARE statement doesn't work
### What happens?

```
PREPARE test_query(text) AS
SELECT count(*) FROM predictions where client = $1;
EXECUTE test_query('example');
```
I get this warning when I run the query and it takes forever:
WARNING:  This query was not fully pushed down to DuckDB because DuckDB returned an error. Query times may be impacted. If you would like to see this query pushed down, please submit a request to https://github.com/paradedb/paradedb/issues with the following context:
Not implemented Error: Prepared statement argument types are not supported, use CAST

With this identical query it works in a few milliseconds:
```
SELECT count(*) FROM predictions where client = 'example';
```

DuckDB itself supports it: https://duckdb.org/docs/api/c/prepared.html

### To Reproduce

Run the query from above

### OS:

Ubuntu

### ParadeDB Version:

v0.8.4

### Are you using ParadeDB Docker, Helm, or the extension(s) standalone?

ParadeDB pg_lakehouse Extension

### Full Name:

Janis

### Affiliation:

Typewise

### Did you include all relevant data sets for reproducing the issue?

Yes

### Did you include the code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configurations (e.g., CPU architecture, PostgreSQL version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by kysshsy**
# Ticket(s) Closed 

- Closes #51 

## How
Pushdown `PREPARE`, `EXECUTE` and `DEALLOCATE` statement to duckdb directly.

### CODE REVIEW

**Comment by kysshsy** [comment]
Note: There is a issue that PostgreSQL will replan the query when certain catalog changes occur, such as changes to the search path or when a table is deleted. In contrast, DuckDB does not replan when the search path is changed. 
If there are two foreign tables in different schemas and the prepared statements do not specify the schemas, it may lead to ambiguity or errors when referencing the tables.
```
create foreign table public.t1(xxx);
create foreign table s1.t1(xxx);

prepare q1 as select * from t1; 
execute q1; (query table public.t1);

set search_path to 's1';

execute q1; (query table public.t1 in duckdb, but it should be s1.t1 in Postgres);

```

**Comment by philippemnoel** [comment]
> Note: There is a issue that PostgreSQL will replan the query when certain catalog changes occur, such as changes to the search path or when a table is deleted. In contrast, DuckDB does not replan when the search path is changed. If there are two foreign tables in different schemas and the prepared statements do not specify the schemas, it may lead to ambiguity or errors when referencing the tables.
> 
> ```
> create foreign table public.t1(xxx);
> create foreign table s1.t1(xxx);
> 
> prepare q1 as select * from t1; 
> execute q1; (query table public.t1);
> 
> set search_path to 's1';
> 
> execute q1; (query table public.t1 in duckdb, but it should be s1.t1 in Postgres);
> ```

Can we force DuckDB to replan via some API call? We want to imitate the Postgres behaviour as much as possible.

**Comment by kysshsy** [comment]
> > Note: There is a issue that PostgreSQL will replan the query when certain catalog changes occur, such as changes to the search path or when a table is deleted. In contrast, DuckDB does not replan when the search path is changed. If there are two foreign tables in different schemas and the prepared statements do not specify the schemas, it may lead to ambiguity or errors when referencing the tables.
> > ```
> > create foreign table public.t1(xxx);
> > create foreign table s1.t1(xxx);
> > 
> > prepare q1 as select * from t1; 
> > execute q1; (query table public.t1);
> > 
> > set search_path to 's1';
> > 
> > execute q1; (query table public.t1 in duckdb, but it should be s1.t1 in Postgres);
> > ```
> 
> Can we force DuckDB to replan via some API call? We want to imitate the Postgres behaviour as much as possible.

Yes.  We could execute `prepare` in Duckdb again. It will override the plan. I will fix it tomorrow.

**Comment by philippemnoel** [comment]
> > > Note: There is a issue that PostgreSQL will replan the query when certain catalog changes occur, such as changes to the search path or when a table is deleted. In contrast, DuckDB does not replan when the search path is changed. If there are two foreign tables in different schemas and the prepared statements do not specify the schemas, it may lead to ambiguity or errors when referencing the tables.
> > > ```
> > > create foreign table public.t1(xxx);
> > > create foreign table s1.t1(xxx);
> > > 
> > > prepare q1 as select * from t1; 
> > > execute q1; (query table public.t1);
> > > 
> > > set search_path to 's1';
> > > 
> > > execute q1; (query table public.t1 in duckdb, but it should be s1.t1 in Postgres);
> > > ```
> > 
> > 
> > Can we force DuckDB to replan via some API call? We want to imitate the Postgres behaviour as much as possible.
> 
> Yes. We could execute `prepare` in Duckdb again. It will override the plan. I will fix it tomorrow.

Sweet. That sounds wonderful. We want to be as "Postgres idiomatic as possible", so whatever the user experience Postgres users expect is is what we should try to replicate.

**Comment by kysshsy** [comment]
> > > > Note: There is a issue that PostgreSQL will replan the query when certain catalog changes occur, such as changes to the search path or when a table is deleted. In contrast, DuckDB does not replan when the search path is changed. If there are two foreign tables in different schemas and the prepared statements do not specify the schemas, it may lead to ambiguity or errors when referencing the tables.
> > > > ```
> > > > create foreign table public.t1(xxx);
> > > > create foreign table s1.t1(xxx);
> > > > 
> > > > prepare q1 as select * from t1; 
> > > > execute q1; (query table public.t1);
> > > > 
> > > > set search_path to 's1';
> > > > 
> > > > execute q1; (query table public.t1 in duckdb, but it should be s1.t1 in Postgres);
> > > > ```
> > > 
> > > 
> > > Can we force DuckDB to replan via some API call? We want to imitate the Postgres behaviour as much as possible.
> > 
> > 
> > Yes. We could execute `prepare` in Duckdb again. It will override the plan. I will fix it tomorrow.
> 
> Sweet. That sounds wonderful. We want to be as "Postgres idiomatic as possible", so whatever the user experience Postgres users expect is is what we should try to replicate.

Yes, I completely agree. I detect changes in the search path and determine whether to replan in DuckDB. It's ready for review. @rebasedming

File: src/hooks/utility/explain.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    let query_relations = get_query_relations(unsafe { (*query).rtable });
+    if unsafe { (*query).commandType } != pg_sys::CmdType::CMD_SELECT
+        || !is_duckdb_query(&query_relations)
+    {
+        return Ok(true);
+    }
+
+    if unsafe { !(*stmt).options.is_null() } {
+        error!("the EXPLAIN options provided are not supported for DuckDB pushdown queries.");
```
Comment (philippemnoel) [inline src/hooks/utility/explain.rs:41]: Am I correct in understanding that this is fixed in your other PR? Or are there other options that won't be supported?

File: src/hooks/utility/prepare.rs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+) -> Result<bool> {
+    unsafe {
+        let prepared_stmt = pg_sys::FetchPreparedStatement((*stmt).name, true);
+        let plan_source = (*prepared_stmt).plansource;
+
+        if plan_source.is_null() || !(*plan_source).fixed_result {
+            return Ok(true);
+        }
+
+        // we need to make Duckdb replan the `PREPARE` statement when search path changed.
```
Comment (philippemnoel) [inline src/hooks/utility/prepare.rs:132]: to match Postgres' default behaviour

File: src/hooks/utility/prepare.rs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        let need_replan = !pg_sys::OverrideSearchPathMatchesCurrent((*plan_source).search_path);
+
+        #[cfg(feature = "pg17")]
+        let need_replan = !pg_sys::SearchPathMatchesCurrentEnvironment((*plan_source).search_path);
+
+        #[cfg(feature = "pg13")]
+        let cached_plan = pg_sys::GetCachedPlan(plan_source, null_mut(), false, null_mut());
+
+        #[cfg(not(feature = "pg13"))]
+        let cached_plan = pg_sys::GetCachedPlan(plan_source, null_mut(), null_mut(), null_mut());
```
Comment (philippemnoel) [inline src/hooks/utility/prepare.rs:143]: Why do you need to call GetCachedPlan twice here? Could you please add a comment to explain?

**Review by philippemnoel — COMMENTED** [review]
this looks good to me! @Weijun-H could you please review this? If this works with you, I'm happy to get it in

**Comment by Weijun-H** [comment]
I will carefully review this PR today or tomorrow.

File: tests/scan.rs [hunk 4]
```diff
@@ -557,3 +558,91 @@ async fn test_executor_hook_search_path(mut conn: PgConnection, tempdir: TempDir
...
+
+    "PREPARE q1 AS SELECT * FROM t1 WHERE boolean_col = $1".execute(&mut conn);
+
+    let result: Vec<(bool,)> = "EXECUTE q1(true)".fetch_collect(&mut conn);
+    assert_eq!(result.len(), 2);
+
+    "SET search_path TO tpch2".execute(&mut conn);
+    let result: Vec<(bool,)> = "EXECUTE q1(true)".fetch_collect(&mut conn);
+    assert_eq!(result.len(), 1);
+
```
Comment (Weijun-H) [inline tests/scan.rs:646]: maybe add a test for `DEALLOCATE `?

**Review by Weijun-H — APPROVED** [review]
LGTM! Thanks @kysshsy

File: src/hooks/utility/prepare.rs (hunk 3 above)
Comment (kysshsy) [reply src/hooks/utility/prepare.rs]: GetCachedPlan interface changed in PG14. so here is a cfg. Add a simple comment

File: src/hooks/utility.rs [hunk 5]
```diff
@@ -16,17 +16,19 @@
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
 #![allow(clippy::too_many_arguments)]
+#![allow(deprecated)]
```
[the remark this answers is not in the document]
Comment (kysshsy) [reply src/hooks/utility.rs:19]: Yes. seem pgrx::HookResult is deprecated. Maybe we could update the code to use the new recommended approach later.

File: tests/scan.rs (hunk 4 above)
Comment (kysshsy) [reply tests/scan.rs:646]: updated.

File: src/hooks/utility/prepare.rs (hunk 2 above)
Comment (kysshsy) [reply src/hooks/utility/prepare.rs]: improve this comment.

**Review by philippemnoel — APPROVED** [review]

**Comment by philippemnoel** [comment]
Thank you <3

**Pull request merged**